File: java/sdk/src/main/java/network/t0/sdk/common/Headers.java
package network.t0.sdk.common;

/**
 * HTTP header names used for request signing and verification.
 *
 * <p>These headers are used to transport the ECDSA signature and related metadata
 * between the provider and T-0 Network.
 */
public final class Headers {

    /**
     * The signature header containing the ECDSA signature in hex format.
     * Value format: "0x" + hex(signature[65])
     */
    public static final String SIGNATURE = "X-Signature";

    /**
     * The timestamp header containing the Unix timestamp in milliseconds.
     * Value format: decimal string representing milliseconds since epoch
     */
    public static final String SIGNATURE_TIMESTAMP = "X-Signature-Timestamp";

    /**
     * The public key header containing the signer's public key in hex format.
     * Value format: "0x" + hex(publicKey[65])
     */
    public static final String PUBLIC_KEY = "X-Public-Key";

    /**
     * The validity window for timestamp verification (in milliseconds).
     * Requests with timestamps outside this window will be rejected.
     */
    public static final long TIMESTAMP_VALIDITY_WINDOW_MS = 60_000; // 60 seconds

    private Headers() {
        // Constants class
    }

    /**
     * Encodes a timestamp as 8-byte little-endian bytes.
     *
     * @param timestampMs the timestamp in milliseconds since epoch
     * @return 8-byte little-endian encoded timestamp
     */
    public static byte[] encodeTimestamp(long timestampMs) {
        byte[] bytes = new byte[8];
        for (int i = 0; i < 8; i++) {
            bytes[i] = (byte) (timestampMs & 0xFF);
            timestampMs >>= 8;
        }
        return bytes;
    }
}

File: csharp/sdk/T0.ProviderSdk/Common/Headers.cs
using System.Buffers.Binary;

namespace T0.ProviderSdk.Common;

/// <summary>
/// HTTP header names used for request signing and verification.
/// </summary>
public static class Headers
{
    public const string Signature = "X-Signature";
    public const string SignatureTimestamp = "X-Signature-Timestamp";
    public const string PublicKey = "X-Public-Key";

    /// <summary>
    /// The validity window for timestamp verification (60 seconds).
    /// </summary>
    public static readonly TimeSpan TimestampValidityWindow = TimeSpan.FromSeconds(60);

    /// <summary>
    /// Encodes a timestamp as 8-byte little-endian bytes.
    /// </summary>
    public static byte[] EncodeTimestamp(long timestampMs)
    {
        var bytes = new byte[8];
        BinaryPrimitives.WriteUInt64LittleEndian(bytes, (ulong)timestampMs);
        return bytes;
    }
}
